Cleanup deep equal tests (fixes #1597)

xrstf opened the pull request:
**What this PR does / why we need it**:

This removes unnecessary diffs and the dependence on the reflect package in favor of the more descriptive equality package. Also, in one place the deep equality was replaced with a simple `bytes.Equal()`, because we are comparing marshalled JSON blobs.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #1597

**Release note**:
```release-note
NONE
```

xrstf committed:
cleanup deep equal tests (fixes #1597)

This removes unnecessary diffs and the dependence on
the reflect package in favor of the more descriptive
equality package.

alvaroaleman commented:
Thanks for the PR @xrstf !

Please keep the markdown code block when adding your release note, otherwise parsing wont work properly ;)

Do you mind adding a little script to the CI that checks for usage of `reflect.DeepEqual` within our code and fails if it finds occurences?

xrstf commented:
> Do you mind adding a little script to the CI that checks for usage of reflect.DeepEqual within our code and fails if it finds occurences?

Sure thing, just by adding a shell script and grepping through the source, exiting with 1 if some usages are found (à la `vertify-addon-version.sh`)?

p0lyn0mial commented:
> Sure thing, just by adding a shell script and grepping through the source, exiting with 1 if some usages are found (à la vertify-addon-version.sh)? 

It would be really awesome to extend an existing one or even write a new "linter" and add it to the `gometalinter` - which is used already in our CI pipeline.

https://github.com/alecthomas/gometalinter#adding-custom-linters

mrIncompetent commented:
We don't use gometalinter in our ci. We use a subset of the linters and call them directly

p0lyn0mial commented:
> We don't use gometalinter in our ci

Why we are not using `gometalinter` directly ? Too much trouble to teach him how to run only a subset ?

xrstf commented:
Does anyone know an existing linter that checks for usage of a given function? Otherwise, sure, I could fork errcheck, call if "funcblacklist" and build it to take a list of forbidden functions.

alvaroaleman commented:
> Sure thing, just by adding a shell script and grepping through the source, exiting with 1 if some usages are found (à la vertify-addon-version.sh)?

That would have been my approach yes but personally I do not care too much about the "how"

p0lyn0mial commented:
if we are not using `gometalinter` then I'm fine with having `a shell script` - whatever works for you.

p0lyn0mial commented:
.. or even checking this at the code review level :D

xrstf committed:
forbid usage of reflect.DeepEqual during builds

xrstf commented:
Example of failed build @ https://drone.loodse.com/kubermatic/kubermatic/2486/8

Excuse my amateurish shell scripting, I'm mostly used to rely on Bash niceties.

alvaroaleman commented:
@xrstf You do have bash when using an ubuntu image - And I much prefer cleaner scripts over POSIX-Conmpatibility ;)

mrIncompetent commented:
gometalinter had some issues in the past with the vendor directory. (And linting vendor hurts.)
Afaik that has been resolved by now. So we could potentially switch back

xrstf commented:
> You do have bash when using an ubuntu image

Yes, but all other shell scripts in `hack/` relied on the base shell and I don't know who's using what shell and didn't want to break if for anyone else. For a small script like this it was acceptable to forgo bash. ;-)

mrIncompetent commented on .drone.yml:
```diff
@@ -31,6 +31,12 @@ pipeline:
+    - ./api/hack/verify-forbidden-functions.sh
+    group: lint
+    image: ubuntu
+    pull: true
```
We generate the drone.yaml from the drone-jsonnet unfortunately. You would need to edit that and regenerate it (README.md should cover that)

xrstf replied:
Oh I must have missed that, sorry!

p0lyn0mial replied:
sorry for the late response but would it be better to add this script to the list of the linters defined in the make file ?

https://github.com/kubermatic/kubermatic/blob/master/api/Makefile#L69

xrstf replied:
The other verify scripts weren't called in the Makefile, which was the reason why I did not add the new one. But since this one is now actually checking *go code* it might make sense in the Makefile.

xrstf committed:
re-generate .drone.yml from jsonnet

xrstf commented:
.drone.yaml is now generated based on the jsonnet file.

xrstf committed:
move verify-forbidden-functions check into the lint Makefile step

It does check go code after all, so it makes sense to have it
be called when someone does "make lint".

p0lyn0mial reviewed (approved):
I am a bit reluctant to let this in because it is almost the same as to say that pointers in general are evil and you dear user are not allowed to use them. There is nothing wrong with `DeepEqual` if you know what objects you are comparing. I think that checking if someone uses `DeepEqual` correctly should be checked during the code review.

I understand that this has strong supportes thus LGTM.

@xrstf  thanks for moving the script to the make file.

xrstf committed:
Merge branch 'master' into cleanup-deep-equal-tests

xrstf committed:
Merge branch 'master' into cleanup-deep-equal-tests
